contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,335 @@
+
+    function _fundIntent(Intent memory intent) internal {
+        address funder = msg.sender;
+        TokenAmount[] memory tokens = intent.reward.tokens;
+
+        // Get vault address from intent
+        address vault = INTENT_SOURCE.intentVaultAddress(intent);
+
+        // Cache tokens length
+        uint256 rewardsLength = tokens.length;
+
+        // Iterate through each token in the reward structure
+        for (uint256 i; i < rewardsLength; ++i) {
+            // Get token address and required amount for current reward
+            address token = tokens[i].token;
+            uint256 amount = tokens[i].amount;
+            uint256 balance = IERC20(token).balanceOf(vault);
+
+            // Only proceed if vault needs more tokens and we have permission to transfer them
+            if (amount > balance) {
+                // Calculate how many more tokens the vault needs to be fully funded
+                uint256 remainingAmount = amount - balance;
+
+                // Check how many tokens this contract is allowed to transfer from funding source
+                uint256 allowance = IERC20(token).allowance(
+                    funder,
+                    address(this)
+                );
+
+                // Check if allowance is sufficient to fund intent
+                if (allowance < remainingAmount) {
+                    revert InsufficientTokenAllowance(
+                        token,
+                        funder,
+                        remainingAmount
+                    );
+                }
+
+                // Transfer tokens from funding source to vault using safe transfer
+                IERC20(token).safeTransferFrom(funder, vault, remainingAmount);
+            }
+        }
+    }
+
+    function _constructIntent(
+        EncodedIntent memory encodedIntent
+    ) internal view returns (Intent memory) {
+        Route memory route = _constructRoute(
+            _randomBytes32(),
+            encodedIntent.sourceChainIndex,
+            encodedIntent.destinationChainIndex,
+            encodedIntent.routeTokenIndex,
+            encodedIntent.routeAmount
+        );
+
+        Reward memory reward = _constructReward(
+            encodedIntent.rewardTokenIndex,
+            encodedIntent.rewardAmount,
+            encodedIntent.expiryDuration
+        );
+
+        return Intent({route: route, reward: reward});
+    }
+
+    function _constructReward(
+        uint8 rewardTokenIndex,
+        uint48 rewardAmount,
+        uint24 expiryDuration
+    ) internal view returns (Reward memory) {
+        TokenAmount memory rewardToken = TokenAmount({
+            token: getTokens()[rewardTokenIndex],
+            amount: rewardAmount
+        });
+
+        TokenAmount[] memory rewardTokens;
+        rewardTokens[0] = rewardToken;
+
+        return
+            Reward({
+                nativeValue: 0,
+                prover: PROVER,
+                creator: msg.sender,
```

Comment:
So we this we can't open intents on someone's behalf?

Reply:
Not for this intent type since passing the `creator` will make the payload over 32 bytes long.

Reply:
yeah that would be a second param, instead of using `msg.sender`. What if solver is using multicall? Should we add `tx.origin` to the mix?

Maybe what u want is a batch method like 
```
publishBatch(address creator, bytes32[] calldata intents) {}
```